src/ellswift.rs

```diff
@@ -0,0 +1,704 @@
+//!
+//! `ElligatorSwift` is described in `https://eprint.iacr.org/2022/759` by
+//! Chavez-Saab, Rodriguez-Henriquez, and Tibouchi. It permits encoding
+//! uniformly chosen public keys as 64-byte arrays which are indistinguishable
+//! from uniformly random arrays.
+//!
+//! Let f be the function from pairs of field elements to point X coordinates,
+//! defined as follows (all operations modulo p = 2^256 - 2^32 - 977)
+//! f(u,t):
+//! - Let C = 0xa2d2ba93507f1df233770c2a797962cc61f6d15da14ecd47d8d27ae1cd5f852,
+//!   a square root of -3.
+//! - If u=0, set u=1 instead.
+//! - If t=0, set t=1 instead.
+//! - If u^3 + t^2 + 7 = 0, multiply t by 2.
+//! - Let X = (u^3 + 7 - t^2) / (2 * t)
+//! - Let Y = (X + t) / (C * u)
+//! - Return the first in [u + 4 * Y^2, (-X/Y - u) / 2, (X/Y - u) / 2] that is an
+//!   X coordinate on the curve (at least one of them is, for any u and t).
+//!
+//! Then an `ElligatorSwift` encoding of x consists of the 32-byte big-endian
+//! encodings of field elements u and t concatenated, where f(u,t) = x.
+//! The encoding algorithm is described in the paper, and effectively picks a
+//! uniformly random pair (u,t) among those which encode x.
+//!
+//! If the Y coordinate is relevant, it is given the same parity as t.
+//!
+//! Changes w.r.t. the paper:
+//! - The u=0, t=0, and u^3+t^2+7=0 conditions result in decoding to the point
+//!   at infinity in the paper. Here they are remapped to finite points.
+//! - The paper uses an additional encoding bit for the parity of y. Here the
+//!   parity of t is used (negating t does not affect the decoded x coordinate,
+//!   so this is possible).
+
+use core::fmt::{self, Display, Formatter};
+use core::ptr;
+use core::str::FromStr;
+
+use ffi::CPtr;
+use secp256k1_sys::types::{c_int, c_uchar, c_void};
+
+use crate::{constants, ffi, from_hex, Error, PublicKey, Secp256k1, SecretKey, Verification};
+
+unsafe extern "C" fn hash_callback<F>(
+    output: *mut c_uchar,
+    x32: *const c_uchar,
+    ell_a64: *const c_uchar,
+    ell_b64: *const c_uchar,
+    hash_func: *mut c_void,
+) -> c_int
+where
+    F: FnMut([u8; 32], [u8; 64], [u8; 64]) -> ElligatorSwiftSharedSecret,
+{
+    let callback: &mut F = &mut *(hash_func as *mut F);
+    let mut x32_array = [0u8; 32];
+    let mut ell_a64_array = [0u8; 64];
+    let mut ell_b64_array = [0u8; 64];
+
+    // Copy the data into Rust slices
+    ptr::copy_nonoverlapping(x32, x32_array.as_mut_c_ptr(), 32);
+    ptr::copy_nonoverlapping(ell_a64, ell_a64_array.as_mut_c_ptr(), 64);
+    ptr::copy_nonoverlapping(ell_b64, ell_b64_array.as_mut_c_ptr(), 64);
+    // Call the hash function that was passed in through the `data` pointer
+    let secret = callback(x32_array, ell_a64_array, ell_b64_array);
+    // Copy the output from a [ElligatorSwiftSharedSecret] into the output pointer
+    ptr::copy_nonoverlapping(secret.0.as_ptr(), output, secret.0.len());
+
+    secret.0.len() as c_int
```

Comment:
In 9b9090b2d777d2dade6f06579fcd0e90a88d8325:

I'm a bit confused by this return value. The docs say that an XDH hash function should return 1 on success, 0 on failure. It looks like we have no way to signal failure, which I think is fine, but it seems we are unconditionally returning `secret.0.len` which will be 64, not 1.